light-client/src/supervisor.rs

```diff
@@ -0,0 +1,273 @@
+        while let Some(primary) = self.peers.primary_mut() {
+            let verdict = match height {
+                None => primary.light_client.verify_to_highest(&mut primary.state),
+                Some(height) => primary
+                    .light_client
+                    .verify_to_target(height, &mut primary.state),
+            };
+
+            match verdict {
+                Ok(light_block) => {
+                    // SAFETY: There must be a latest trusted state otherwise verification would have failed.
+                    let trusted_state = primary
+                        .state
+                        .light_store
+                        .latest(VerifiedStatus::Verified)
+                        .unwrap();
+
+                    let outcome = self.detect_forks(&light_block, &trusted_state)?;
+
+                    match outcome {
+                        Some(forks) => {
+                            let mut forked = Vec::with_capacity(forks.len());
+
+                            for fork in forks {
+                                match fork {
+                                    Fork::Forked(block) => {
+                                        self.report_evidence(&block);
+                                        forked.push(block.provider);
+                                    }
+                                    Fork::Faulty(block, _error) => {
+                                        self.peers.remove_secondary(&block.provider);
+                                        // TODO: Log/record the error
+                                    }
+                                }
+                            }
+
+                            if !forked.is_empty() {
+                                // Fork detected, exiting
+                                bail!(ErrorKind::ForkDetected(forked))
+                            }
+                        }
+                        None => {
+                            // No fork detected, exiting
+                            // TODO: Send to relayer, maybe the run method does this?
+                            return Ok(light_block);
+                        }
+                    }
+                }
+                // Verification failed
+                Err(_err) => {
+                    // Swap primary, and continue with new primary, if any.
+                    self.peers.swap_primary()?;
+                    // TODO: Log/record error
+                    continue;
+                }
+            }
+        }
+
+        bail!(ErrorKind::NoValidPeerLeft)
+    }
+
+    fn report_evidence(&mut self, _light_block: &LightBlock) {
+        ()
+    }
+
+    #[pre(self.peers.primary().is_some())]
+    fn detect_forks(
+        &mut self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+    ) -> Result<Option<Vec<Fork>>, Error> {
+        if self.peers.secondaries().is_empty() {
+            return Ok(None);
```

Comment:
this seems very insecure to me.. even though we rely on 1 correct node assumption, my personal opinion is that we need at least 1 witness. we can't assume primary will always be correct.